Framework/Algorithms/src/EstimateResolutionDiffraction.cpp

```diff
@@ -180,40 +209,55 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
     }
 
     // Get the distance from detector to source
-    double l2 = spectrumInfo.l2(i);
+    const double l2 = spectrumInfo.l2(i);
 
     // Calculate T
-    double centraltof = (l1 + l2) / m_centreVelocity;
+    const double centraltof = (l1 + l2) / m_centreVelocity;
 
     // Angle
-    double twotheta =
+    const double twotheta =
         spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
-    double theta = 0.5 * twotheta;
+    const double theta = 0.5 * twotheta;
 
-    double solidangle = 0.0;
-    for (const auto detID : m_inputWS->getSpectrum(i).getDetectorIDs()) {
-      const auto index = detectorInfo.indexOf(detID);
-      if (!detectorInfo.isMasked(index))
-        solidangle += detectorInfo.detector(index).solidAngle(samplepos);
+    double deltatheta = 0.;
+    if (m_divergenceWS) {
+      deltatheta = m_divergenceWS->readY(i)[0];
+    } else {
+      double solidangle = 0.0;
+
+      for (const auto detID : m_inputWS->getSpectrum(i).getDetectorIDs()) {
+        const auto index = detectorInfo.indexOf(detID);
```

Comment:
Similar to above, use 
```cpp
for (const auto &index : spectrumInfo.spectrumDefinition(i))
  if (!detectorInfo.isMasked(index.first))
    solidangle += componentInfo.solidAngle(index.first, samplepos);
```
Again, currently this does not support scanning, but your solution doesn't either and this way it will be a breeze to add support later (just remove `.first`).